### PR TITLE
PARQUET-2211: [C++] Print ColumnMetaData.encoding_stats field

### PR DESCRIPTION
**Pull request by wgtmac**

### CODE REVIEW

**Comment by wgtmac** [comment]
@pitrou Can you please take a look? Thanks!

**Review by emkornfield — APPROVED** [review]

**Comment by wgtmac** [comment]
@emkornfield Thanks for the review! Do you have the chance to merge it if possible?

File: cpp/src/parquet/printer.cc [hunk 1]
```diff
@@ -39,6 +39,25 @@ namespace parquet {
...
+namespace {
+
+void PrintPageEncodingStats(std::ostream& stream,
+                            const std::vector<PageEncodingStats>& encoding_stats) {
+  for (size_t i = 0; i < encoding_stats.size(); ++i) {
+    const auto& encoding = encoding_stats.at(i);
+    stream << EncodingToString(encoding.encoding);
+    if (encoding.page_type == parquet::PageType::DICTIONARY_PAGE) {
+      // Explicitly tell if this encoding comes from a dictionary page
+      stream << "(DICT_PAGE)";
```
Comment (pitrou) [inline cpp/src/parquet/printer.cc:51]: Can you explain what this means? Can a dictionary encoding come from a non-dict page, or vice-versa?

File: cpp/src/parquet/reader_test.cc [hunk 2]
```diff
@@ -547,12 +547,12 @@ Column 1: b (INT32)
 --- Rows: 3 ---
 Column 0
   Values: 18  Statistics Not Set
-  Compression: SNAPPY, Encodings: RLE PLAIN_DICTIONARY
+  Compression: SNAPPY, Encodings: PLAIN_DICTIONARY(DICT_PAGE) PLAIN_DICTIONARY
```
Comment (pitrou) [inline cpp/src/parquet/reader_test.cc:550]: This is weird. The same data was RLE-encoded before this PR and suddenly it is plain-dict encoded?

File: cpp/src/parquet/reader_test.cc [hunk 3]
```diff
@@ -547,12 +547,12 @@ Column 1: b (INT32)
...
   Uncompressed Size: 103, Compressed Size: 104
 Column 1
   Values: 3, Null Values: 0, Distinct Values: 0
   Max: 1, Min: 1
-  Compression: SNAPPY, Encodings: BIT_PACKED PLAIN_DICTIONARY
+  Compression: SNAPPY, Encodings: PLAIN_DICTIONARY(DICT_PAGE) PLAIN_DICTIONARY
```
Comment (pitrou) [inline cpp/src/parquet/reader_test.cc:555]: Same question about this one.

File: cpp/src/parquet/reader_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/reader_test.cc:550]: Perhaps RLE is how dictionary indices are encoded... but then why is this information not displayed anymore?

Comment (pitrou) [reply cpp/src/parquet/reader_test.cc:550]: (but then it would have been RLE_DICTIONARY?)

File: cpp/src/parquet/printer.cc (hunk 1 above)
Comment (wgtmac) [reply cpp/src/parquet/printer.cc:51]: The main idea is to tell this encoding comes from the dictionary page. IIUC, both dictionary page and data page use PLAIN_DICTIONARY when dictionary encoding is applied in the Parquet 1.0. While in Parquet 2.0, dictionary page uses PLAIN and data page uses RLE_DICTIONARY. So it is difficult to tell where the PLAIN_DICTIONARY or PLAIN encoding comes from.  Please check this for detail:　https://github.com/apache/parquet-format/blob/master/Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8

File: cpp/src/parquet/reader_test.cc (hunk 2 above)
Comment (wgtmac) [reply cpp/src/parquet/reader_test.cc:550]: The RLE encoding was used for def/rep levels. After this change, it only prints encoding of data. This is in sync with what parquet-mr does: https://github.com/apache/parquet-mr/blob/c00479d538353348726cf78da835078024161e55/parquet-cli/src/main/java/org/apache/parquet/cli/commands/ParquetMetadataCommand.java#L163

File: cpp/src/parquet/reader_test.cc (hunk 3 above)
Comment (wgtmac) [reply cpp/src/parquet/reader_test.cc:555]: Same here, BIT_PACKED was used for def/rep levels.

File: cpp/src/parquet/reader_test.cc (hunk 2 above)
Comment (wgtmac) [reply cpp/src/parquet/reader_test.cc:550]: So the main idea is that we should always prefer encoding_stats to encoding if encoding_stats exists when printing encoding metadata.

Comment (pitrou) [reply cpp/src/parquet/reader_test.cc:550]: I see, thanks.

**Review by pitrou — APPROVED** [review]
Thanks for this. LGTM but Parquet encodings are painfully complicated.

**Pull request merged**